Displaying amazon network routers

Ladas opened the pull request:
Depends on:
- [x] https://github.com/ManageIQ/manageiq/pull/16030
 
Displaying amazon network routers

Partially implements:
https://bugzilla.redhat.com/show_bug.cgi?id=1488135

Before:
we were not collecting AWS routers

After:
![screenshot from 2017-09-25 16-45-11](https://user-images.githubusercontent.com/1737058/30853972-93a2cdf6-a2b0-11e7-820d-c9ecbec86060.png)
![screenshot from 2017-09-25 15-05-36](https://user-images.githubusercontent.com/1737058/30853982-9b21390a-a2b0-11e7-997e-3e009f9a605c.png)

Ladas commented:
cc @bronaghs routers RFE

@romanblanco can you check if the supports checks are correct? Not sure why they were missing. Also I am a bit confused by list actions, seems like we can't do support checks there? I see that in other models all list actions are commented out, I used the same comment as @ZitaNemeckova with https://github.com/ManageIQ/manageiq/pull/12551 . Not sure if that is a part of a bigger plan to do support checks in list actions.

durandom commented on app/helpers/application_helper/toolbar/network_routers_center.rb:
```diff
@@ -15,41 +15,43 @@ class ApplicationHelper::Toolbar::NetworkRoutersCenter < ApplicationHelper::Tool
-            :enabled   => false,
-            :onwhen    => '1+'),
+          # TODO: Uncomment until cross controllers show_list issue fully in place/also supports? must be in place
+          # https://github.com/ManageIQ/manageiq/pull/12551
```
this PR has been long merged.
Is there any javascript logic, that enables/disables those buttons?

Ladas replied:
Yes it was, but the code is commented like this in most of the table toolbars. I don't think we can uncomment it until we solve the ```supports?``` there, since now I am getting exceptions in unsupported providers, because of missing methods. (AFAIK there is no JS logic for table level buttons&support mixin)

durandom replied:
ok, I dont really understand whats needed from the support mixin here. Maybe @ZitaNemeckova and @romanblanco could chime in, if this is a shortcoming in the supports_feature_mixin

Ladas commented:
@miq-bot assign @martinpovolny

Ladas commented:
@miq-bot add_label enhancement

ohadlevy commented on app/helpers/application_helper/toolbar/network_routers_center.rb:
```diff
@@ -15,41 +15,42 @@ class ApplicationHelper::Toolbar::NetworkRoutersCenter < ApplicationHelper::Tool
-            :confirm   => N_('Warning: The selected Routers and ALL of their components will be removed!'),
-            :enabled   => false,
-            :onwhen    => '1+'),
+          # TODO: Uncomment when support? for list actions works on some degree
```
why commit this?

Ladas replied:
it's like that in other files. Once we deal with supports? these actions are working

Ladas replied:
although in other files it references non related PR, from what the UI guys told me

ohadlevy replied:
I'm not sure I follow, what do you mean by deal with support? bottom line - I can't see how adding in commented out code is acceptable?

martinpovolny replied:
@Ladas: I know that you are referring to a situation where someone (bet it was not me) commented out some code.
As I see it, I'd rather delete the code.

On "Once we deal with supports? these actions are working" I don't think so or I don't see how. Please, explain.

I have explained a couple of times why we cannot have complex logic around enabling/disabling, showing/hiding toolbar buttons __in list views__ so I am not going to repeat it here again.

I think that the general approach should be:

 1. having only the __basic actions__ in the __list view__ toolbars and 
 1. having the __full range__ of functions in the toolbars on the __details screen__.

Ladas replied:
Ah, so the actions are working now, since OpenStack was the only provider with Routers. I am adding AWS routers, so the actions will fail there.

Right, as I mentioned, it can be as as simple as assert_support line in each action, to move the check.

Ok, I will delete the commented code. My first try is always consistence. ;-)

ohadlevy commented on app/helpers/network_router_helper/textual_summary.rb:
```diff
@@ -24,10 +24,43 @@ def textual_type
+    return nil if @record.main_route_table.nil?
+
+    if @record.main_route_table
+      message = "(Subnets not explicitely assigned to a route table are assigned to this main route table, for this VPC)"
```
i18n?

Ladas replied:
gah, thanks

Ladas commented:
@ohadlevy I've added specs long long time ago https://github.com/Ladas/manageiq-ui-classic/blob/679cb360fe7e7b8f180cef1a8e27532ac18b23f3/spec/controllers/network_router_controller_spec.rb#L2 leading to https://github.com/Ladas/manageiq-ui-classic/blob/09b1867835301a066fb884d09bcda3a9072110cb/spec/shared/controllers/shared_examples_for_network_router_controller.rb#L3-L2

ohadlevy commented:
@Ladas regarding tests, I can't understand how old tests check your new functionality? clearly new tests needs to be written?

ohadlevy commented:
@Ladas also note the test failures.

Ladas commented:
@ohadlevy the spec failure is a badly written spec ```it { is_expected.to eq([@vm1, @vm2]) }``` it needs to do match_array, cause the order is not guaranteed, so it can randomly fail now.

The current specs are a basic integration specs using the AWS Router model, we do not really test more details of how we render the detail page anywhere.

martinpovolny commented:
The failing specs are due to a master breakage caused here: ManageIQ/manageiq#15964

martinpovolny commented:
> The current specs are a basic integration specs using the AWS Router model, we do not really test more details of how we render the detail page anywhere.

@Ladas : does that mean that this is a wip, because it lacks the tests? Or is it an excuse for not writing the tests (and later complaining when someone breaks this) ;-) ?

Ladas commented:
@martinpovolny @ohadlevy no what I mean, I added specs for NetworkRouter for every provider while ago. Now we are actually collecting those for AWS, but we were able to display the AWS routers before.

The specs are https://github.com/Ladas/manageiq-ui-classic/blob/09b1867835301a066fb884d09bcda3a9072110cb/spec/shared/controllers/shared_examples_for_network_router_controller.rb#L3-L2

So, the same basic specs we use for most of the controllers. So if you want more detailed/other specs, let me know what is currently considered as good specs in UI classic. (before we had almost no specs, I added at least these basic specs, what is the next level?)

martinpovolny commented:
@Ladas : I am currently deep in the GTLs figuring out how to test it the best way. So let's skip that for this PR.

BUT: you are adding new stuff to: `app/helpers/network_router_helper/textual_summary.rb`. Is there a spec that runs that code and at least makes sure it renders? That's all I want in this PR.

Thx!

Ladas commented:
@martinpovolny yes, that is in this spec https://github.com/Ladas/manageiq-ui-classic/blob/09b1867835301a066fb884d09bcda3a9072110cb/spec/shared/controllers/shared_examples_for_network_router_controller.rb#L38

For having 100% coverage, we need to have the right attributes in the factories, which are in AWS repo.

martinpovolny commented:
> For having 100% coverage, we need to have the right attributes in the factories, which are in AWS repo.

What about adding a factory (possibly one that is based on the one from the AWS repo) in this repo? Does that make sense?

Ladas commented:
@martinpovolny originally all factories were in https://github.com/Ladas/manageiq/blob/22284143d374a1818c6db232d6800df11db52b34/spec/factories/network_router.rb

We are slowly moving them to provider repos e.g. https://github.com/Ladas/manageiq-providers-amazon/blob/616ecc680196711018a3940381a1dd1b6ca30518/spec/factories/floating_ip.rb , network_router is not moved yet

Hm, not sure, I think the plan was to move all 'amazon related' objects under the amazon repo. But when the integration specs live in the UI, maybe we could touch the factories here.

martinpovolny commented:
> Hm, not sure, I think the plan was to move all 'amazon related' objects under the amazon repo. But when the integration specs live in the UI, maybe we could touch the factories here.

We need to test the UI somewhere right? So we need the factories. In the backend they don't have the same need so what happens from time to time, is that they actually remove the factories.

And you are right, the UI tests have the role of integration tests in this case.

Please, create a factory with the necessary values, extend the existing or add a new tests and lets merge this! That should be a 20 minutes' task for you.

Ladas commented:
@martinpovolny hum, I think we use specific factories not, for the generic tests. Let me see what I can do. (need to finish other stuff first)

martinpovolny commented:
Need to add the list of BZs here.
Need to add test coverage for this.

Ladas commented:
@martinpovolny added list of BZs this fixes

We need specs for buttons (that every button is covered by supports?)

We have specs for textual pages, but we need to enhance them to use more data combination, so starting with various factories in provider repos, that can be then used for testing that the textual pages don't break. Not sure how to generalize that yet.Maybe we could even test that in providers themselves (I think I started that with the shared specs) and somehow use the data from the VCRs (so we don't have to manually mock thousands of data combinations)

simaishi commented:
Gaprindashvili backport details:

```
$ git log -1
commit 01db70a948226b211343712986e5113e0cb7e4d5
Author: Martin Povolny <mpovolny@redhat.com>
Date:   Tue Jan 30 09:38:40 2018 +0100

    Merge pull request #2234 from Ladas/displaying_amazon_network_routers
    
    Displaying amazon network routers
    (cherry picked from commit 916d743d48d13cb8934532033cf7dc55d8b1b55d)
    
    Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1540743
    Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1540744
    Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1540745
    Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1540746
```

Ladas commented:
@simaishi lets change this to Fine/no, this was a RFE

simaishi commented:
@Ladas Thanks.

@h-kataria I think you added `fine/yes` for this [5.8.z BZ](https://bugzilla.redhat.com/show_bug.cgi?id=1544488)? Not sure if you can create a Fine PR, just taking what you need for the BZ?

h-kataria commented:
@simaishi i added fine/yes based upon version on BZ, i don't mind changing it to fine/no

Ladas commented:
Ok, let's change this to Fine/no, it would be nontrivial amount of work to have the backend for Fine

@miq-bot add_label fine/no
@miq-bot remove_label fine/yes
